### PR TITLE
Add more mac messages to setup.py

### LINKED ISSUE #105060: extra information messages for mac in setup.py would help.
### 🐛 Describe the bug

I am following [https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md] (https://github.com/pytorch/pytorch/blob/main/CONTRIBUTING.md) on an M1 mac, to setup the development source build.

I can successfully run `python setup.py develop` but it does not seem to have done anything. Looking at the code in setup.py I would have expected the IS_DARWIN to be set and also CMAKE_OSX_ARCHITECTURES to be set appropriately. For me these are not set. 

to get this to work I used 

USE_DISTRIBUTED=0 CMAKE_OSX_ARCHITECTURES=arm64 MACOSX_DEPLOYMENT_TARGET=11.0 USE_MKLDNN=OFF USE_QNNPACK=OFF WERROR=1 BUILD_TEST=OFF USE_PYTORCH_METAL=1 python setup.py develop

I would have liked more information messages to be put out by the setup.py. I suggest the following code

```
 # Cross-compile for M1
    if IS_DARWIN:
        report('-- detected IS_DARWIN')
        macos_target_arch = os.getenv('CMAKE_OSX_ARCHITECTURES', '')

        if macos_target_arch in ['arm64', 'x86_64']:
            report('-- detected macos target architecture ' + macos_target_arch)
            macos_sysroot_path = os.getenv('CMAKE_OSX_SYSROOT')
            if macos_sysroot_path is None:
                macos_sysroot_path = subprocess.check_output([
                    'xcrun', '--show-sdk-path', '--sdk', 'macosx'
                ]).decode('utf-8').strip()
            extra_compile_args += ['-arch', macos_target_arch, '-isysroot', macos_sysroot_path]
            extra_link_args += ['-arch', macos_target_arch]
        elif macos_target_arch == '':
            report('-- no target architecture found. Please ensure CMAKE_OSX_ARCHITECTURES environment variable is set.')
        else:
            report('-- unrecognised macos target architecture ' + macos_target_arch + '. CMAKE_OSX_ARCHITECTURES environment variable does not contain a valid target architecture name.')
```

### Versions

# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
--2023-07-12 15:27:01--  https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 2606:50c0:8002::154, 2606:50c0:8001::154, 2606:50c0:8003::154, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|2606:50c0:8002::154|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 21653 (21K) [text/plain]
Saving to: ‘collect_env.py’

collect_env.py             100%[=======================================>]  21.15K  --.-KB/s    in 0s      

2023-07-12 15:27:02 (42.4 MB/s) - ‘collect_env.py’ saved [21653/21653]

zsh: command not found: #
Collecting environment information...
PyTorch version: 2.1.0a0+gitc03558f
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.4.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.10.12 | packaged by conda-forge | (main, Jun 23 2023, 22:41:52) [Clang 15.0.7 ] (64-bit runtime)
Python platform: macOS-13.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==0.960
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] torch==2.1.0a0+gitf353d17
[pip3] torchgen==0.0.1
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0a0+gitc03558f           dev_0    <develop>
[conda] torchgen                  0.0.1                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by davidradl**
Fixes #105060

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -569,6 +569,7 @@ def run(self):
         setuptools.command.build_ext.build_ext.run(self)
 
         if IS_DARWIN and package_type != 'conda':
+            report('-- Building libiomp')
```
Comment (malfet) [inline setup.py:572]: I think this is misleading, as it's not building libiomp, but bundles it into the package

File: setup.py [hunk 2]
```diff
@@ -887,16 +888,24 @@ def configure_extension_build():
 
     # Cross-compile for M1
     if IS_DARWIN:
+        report('-- detected IS_DARWIN')
```
Comment (malfet) [inline setup.py:891]: This information is already printed in Cmake summary
```suggestion
```

File: setup.py [hunk 3]
```diff
@@ -887,16 +888,24 @@ def configure_extension_build():
...
             if macos_sysroot_path is None:
                 macos_sysroot_path = subprocess.check_output([
                     'xcrun', '--show-sdk-path', '--sdk', 'macosx'
                 ]).decode('utf-8').strip()
             extra_compile_args += ['-arch', macos_target_arch, '-isysroot', macos_sysroot_path]
             extra_link_args += ['-arch', macos_target_arch]
-
+        elif macos_target_arch == '':
+            # We are adding no extra compile and link args. Can this ever be valid?
+            report('-- no target architecture found.')
```
Comment (malfet) [inline setup.py:905]: This is totally fine, i.e. path taken for the native builds, isn't it?

File: setup.py [hunk 4]
```diff
@@ -887,16 +888,24 @@ def configure_extension_build():
...
+        else:
+            # This could be a watch architecture like armv7k-arm64-32. We are adding no extra compile and link args. is this valid?
+            report('-- macos target architecture is ' + macos_target_arch + '.')
```
Comment (malfet) [inline setup.py:908]: Please remove the comment, but printing a warning is fine here, though I would prefer to use term [Darwin](https://en.wikipedia.org/wiki/Darwin_(operating_system)) as [CMAKE_OSX_ARCHITECTURES](https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_ARCHITECTURES.html) applies to both macOS, iOS etc

File: setup.py [hunk 5]
```diff
@@ -887,16 +888,24 @@ def configure_extension_build():
 
     # Cross-compile for M1
     if IS_DARWIN:
+        report('-- detected IS_DARWIN')
         macos_target_arch = os.getenv('CMAKE_OSX_ARCHITECTURES', '')
+
         if macos_target_arch in ['arm64', 'x86_64']:
+            report('-- detected macos target architecture ' + macos_target_arch)
```
Comment (malfet) [inline setup.py:895]:
```suggestion
            report(f"-- Compiling for  macOS target architecture {macos_target_arch}")
```

**Comment by malfet** [comment]
Also, I think `Fixes` message references this PR, rather than issue

**Comment by davidradl** [comment]
> Also, I think `Fixes` message references this PR, rather than issue

good spot - changed

File: setup.py (hunk 3 above)
Comment (davidradl) [reply setup.py:905]: @malfet Thanks tor you comments - can I check what you mean by a native build. Do you mean native as in the ATen native folder devices? If so would this mean `mps` for Apple and we are saying we do not need any extra args for `mps`. I will make the changes assuming this is the case.

File: setup.py (hunk 1 above)
Comment (davidradl) [reply setup.py]: amended

File: setup.py (hunk 2 above)
Comment (davidradl) [reply setup.py]: removed

File: setup.py (hunk 4 above)
Comment (davidradl) [reply setup.py]: changed

File: setup.py (hunk 3 above)
Comment (davidradl) [reply setup.py]: changed

File: setup.py (hunk 5 above)
Comment (davidradl) [reply setup.py]: I changed to `building` as there is `linking` as well as `compiling` and `MacOS` to `Darwin`.

File: setup.py [hunk 6]
```diff
@@ -888,15 +889,20 @@ def configure_extension_build():
...
                 macos_sysroot_path = subprocess.check_output([
                     'xcrun', '--show-sdk-path', '--sdk', 'macosx'
                 ]).decode('utf-8').strip()
             extra_compile_args += ['-arch', macos_target_arch, '-isysroot', macos_sysroot_path]
             extra_link_args += ['-arch', macos_target_arch]
-
+        elif macos_target_arch == '':
+            report('-- No target architecture found.')
+        else:
+            report(f"-- Found Darwin target architecture {macos_target_arch}")
```
Comment (malfet) [inline setup.py:905]: This should be an error or a warning, right? I.e. `macos_target_arch` which is not either arm64 or x86_64 is ignored, isn't it?

File: setup.py [hunk 7]
```diff
@@ -888,15 +889,20 @@ def configure_extension_build():
...
             macos_sysroot_path = os.getenv('CMAKE_OSX_SYSROOT')
             if macos_sysroot_path is None:
                 macos_sysroot_path = subprocess.check_output([
                     'xcrun', '--show-sdk-path', '--sdk', 'macosx'
                 ]).decode('utf-8').strip()
             extra_compile_args += ['-arch', macos_target_arch, '-isysroot', macos_sysroot_path]
             extra_link_args += ['-arch', macos_target_arch]
-
+        elif macos_target_arch == '':
+            report('-- No target architecture found.')
```
Comment (malfet) [inline setup.py:903]:
```suggestion
            report('-- No target Darwin architecture specified, running default build.')
```

**Comment by kit1980** [comment]
@davidradl I wonder if you're still interested in this PR?

**Pull request closed without merging**